File: backend/app/schemas/pos.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Any, Literal
import uuid

from pydantic import ConfigDict, Field, model_validator

from app.schemas import BaseSchema
from app.schemas.pricing import PriceOverrideIntent, PricingChannel


class OpenShiftRequest(BaseSchema):
    location_id: uuid.UUID
    opening_cash: Decimal = Field(default=Decimal("0"), ge=0)
    shift_type: Literal["staff_cashier", "operational_cashless"] = "staff_cashier"
    idempotency_key: str | None = Field(default=None, min_length=8, max_length=100)


class CashDenominationCount(BaseSchema):
    denomination: Decimal = Field(gt=0)
    quantity: int = Field(ge=0, le=10000)


ShiftVarianceReason = Literal[
    "count_short",
    "count_over",
    "change_error",
    "cash_movement",
    "other",
]


class CloseShiftRequest(BaseSchema):
    closing_cash: Decimal = Field(ge=0)
    reason_code: ShiftVarianceReason | None = None
    note: str | None = Field(default=None, max_length=500)
    cash_count: list[CashDenominationCount] = Field(default_factory=list, max_length=20)
    expected_version: int | None = Field(default=None, ge=1)
    idempotency_key: str | None = Field(default=None, min_length=8, max_length=100)
    approval_token: str | None = None

    @model_validator(mode="after")
    def validate_cash_count(self) -> "CloseShiftRequest":
        if self.cash_count:
            counted = sum(
                (Decimal(item.denomination) * item.quantity for item in self.cash_count),
                Decimal("0"),
            ).quantize(Decimal("0.01"))
            if counted != Decimal(self.closing_cash).quantize(Decimal("0.01")):
                raise ValueError("Cash denomination total must equal closing cash")
        return self


CashMovementType = Literal["cash_in", "cash_out"]
CashMovementReason = Literal[
    "change_fund",
    "cash_drop",
    "petty_cash",
    "supplier_payment",
    "correction",
    "other",
]


class CashMovementCreateRequest(BaseSchema):
    movement_type: CashMovementType
    amount: Decimal = Field(gt=0)
    reason_code: CashMovementReason
    reason: str = Field(min_length=3, max_length=500)
    expected_shift_version: int = Field(ge=1)
    idempotency_key: str = Field(min_length=8, max_length=100)
    approval_token: str | None = None


class ShiftRead(BaseSchema):
    id: uuid.UUID
    shift_number: str
    shift_type: str = "staff_cashier"
    version: int = 1
    status: str
    branch_id: uuid.UUID
    location_id: uuid.UUID
    user_id: uuid.UUID
    opened_at: datetime
    closed_at: datetime | None = None
    opening_cash: Decimal
    closing_cash: Decimal | None = None
    expected_cash: Decimal | None = None
    cash_difference: Decimal | None = None
    total_sales: Decimal
    total_orders: int
    total_voids: int
    close_reason_code: str | None = None
    cash_count_json: list[dict[str, Any]] | None = None
    opened_device_id: uuid.UUID | None = None
    opened_device_code: str | None = None
    closed_device_id: uuid.UUID | None = None
    closed_device_code: str | None = None
    closed_by_user_id: uuid.UUID | None = None
    close_snapshot_json: dict[str, Any] | None = None

    model_config = ConfigDict(from_attributes=True)


class CashMovementRead(BaseSchema):
    id: uuid.UUID
    shift_id: uuid.UUID
    movement_type: str
    amount: Decimal
    reason_code: str
    reason: str
    status: str
    shift_version: int
    requester_id: uuid.UUID
    approver_id: uuid.UUID | None = None
    device_code: str | None = None
    journal_entry_id: uuid.UUID | None = None
    posted_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CartItem(BaseSchema):
    product_id: uuid.UUID
    variant_id: uuid.UUID | None = None
    qty: Decimal
    unit_price: Decimal
    original_price: Decimal
    discount_amount: Decimal = Decimal("0")
    discount_type: str = "amount"
    vat_type: str
    vat_rate: Decimal = Decimal("7")
    expected_price_version: str | None = None
    price_override: PriceOverrideIntent | None = None


class PaymentCreateRequest(BaseSchema):
    payment_method: str
    amount: Decimal
    reference_no: str | None = None


class CreateSaleRequest(BaseSchema):
    shift_id: uuid.UUID
    location_id: uuid.UUID
    items: list[CartItem] = Field(default_factory=list)
    discount_amount: Decimal = Decimal("0")
    discount_type: str = "amount"
    payment_method: str
    payments: list[PaymentCreateRequest] = Field(default_factory=list)
    payment_reference: str | None = None
    paid_amount: Decimal
    customer_name: str | None = None
    customer_phone: str | None = None
    customer_tax_id: str | None = None
    customer_id: uuid.UUID | None = None
    note: str | None = None
    is_offline: bool = False
    client_order_id: str | None = None
    approval_token: str | None = None
    price_override_approval_token: str | None = None
    pricing_quote_id: uuid.UUID | None = None
    pricing_calculation_hash: str | None = Field(default=None, min_length=64, max_length=64)
    channel: PricingChannel = "pos"
    currency: str = Field(default="THB", min_length=3, max_length=3)
    cart_version: int = Field(default=1, ge=1)
    source_hold_draft_id: uuid.UUID | None = None
    source_hold_draft_version: int | None = Field(default=None, ge=1)


class PaymentRead(BaseSchema):
    id: uuid.UUID
    order_id: uuid.UUID
    payment_method: str
    amount: Decimal
    reference_no: str | None = None
    original_payment_id: uuid.UUID | None = None
    currency: str = "THB"
    provider_name: str | None = None
    provider_payment_ref: str | None = None
    settlement_state: str = "unknown"
    refund_operation_id: uuid.UUID | None = None
    provider_refund_state: str | None = None
    paid_at: datetime

    model_config = ConfigDict(from_attributes=True)


class SaleOrderItemRead(BaseSchema):
    id: uuid.UUID
    product_id: uuid.UUID
    variant_id: uuid.UUID | None = None
    product_name: str
    variant_name: str | None = None
    sku: str
    unit_code: str | None = None
    qty: Decimal
    unit_price: Decimal
    original_price: Decimal
    discount_amount: Decimal
    vat_type: str
    vat_rate: Decimal
    vat_amount: Decimal
    subtotal: Decimal
    refunded_qty: Decimal = Decimal("0")
    refunded_amount: Decimal = Decimal("0")
    price_source: str | None = None
    price_list_id: uuid.UUID | None = None
    price_list_version: int | None = None
    price_version: str | None = None
    price_snapshot: dict | None = None
    order_discount_share: Decimal = Decimal("0")
    line_total: Decimal = Decimal("0")
    price_override_applied: bool = False
    price_override_reason: str | None = None

    model_config = ConfigDict(from_attributes=True)


class SaleOrderRead(BaseSchema):
    id: uuid.UUID
    order_number: str
    status: str
    branch_id: uuid.UUID
    location_id: uuid.UUID
    shift_id: uuid.UUID
    user_id: uuid.UUID
    customer_name: str | None = None
    customer_phone: str | None = None
    customer_tax_id: str | None = None
    subtotal: Decimal
    discount_amount: Decimal
    vat_amount: Decimal
    total_amount: Decimal
    refund_amount: Decimal = Decimal("0")
    paid_amount: Decimal
    change_amount: Decimal
    is_offline: bool
    synced_at: datetime | None = None
    recipe_stock_status: str | None = None
    recipe_stock_warnings: str | None = None
    recipe_stock_posted_at: datetime | None = None
    recipe_stock_reversed_at: datetime | None = None
    note: str | None = None
    pricing_quote_id: uuid.UUID | None = None
    pricing_request_hash: str | None = None
    pricing_calculation_hash: str | None = None
    pricing_calculation_version: str | None = None
    pricing_context: dict | None = None
    pricing_snapshot: dict | None = None
    row_version: int = 1
    created_at: datetime
    items: list[SaleOrderItemRead] = Field(default_factory=list)
    payments: list[PaymentRead] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)


class VoidRequest(BaseSchema):
    void_reason: str
    approval_token: str | None = None


class RefundRequest(BaseSchema):
    refund_reason: str
    approval_token: str | None = None


class PartialRefundItemRequest(BaseSchema):
    order_item_id: uuid.UUID
    qty: Decimal


class PartialRefundRequest(BaseSchema):
    refund_reason: str
    items: list[PartialRefundItemRequest] = Field(default_factory=list)
    approval_token: str | None = None


RefundReasonCode = Literal[
    "customer_request",
    "wrong_item",
    "quality_issue",
    "duplicate_charge",
    "payment_error",
    "other",
]


class RefundQuoteItemRequest(BaseSchema):
    order_item_id: uuid.UUID
    qty: Decimal = Field(gt=0)


class RefundQuoteCreateRequest(BaseSchema):
    order_id: uuid.UUID
    shift_id: uuid.UUID
    items: list[RefundQuoteItemRequest] = Field(default_factory=list, max_length=100)
    reason_code: RefundReasonCode
    reason_note: str | None = Field(default=None, max_length=500)
    stock_disposition: Literal["none", "sellable"] = "none"
    currency: str = Field(default="THB", min_length=3, max_length=3)
    idempotency_key: str = Field(min_length=8, max_length=100)

    @model_validator(mode="after")
    def validate_other_reason(self) -> "RefundQuoteCreateRequest":
        if self.reason_code == "other" and len((self.reason_note or "").strip()) < 3:
            raise ValueError("reason_note is required when reason_code is other")
        return self


class RefundExecuteRequest(BaseSchema):
    quote_id: uuid.UUID
    quote_hash: str = Field(min_length=64, max_length=64)
    order_id: uuid.UUID
    expected_order_version: int = Field(ge=1)
    total_amount: Decimal = Field(gt=0)
    reason_code: RefundReasonCode
    reason_note: str | None = Field(default=None, max_length=500)
    stock_disposition: Literal["none", "sellable"] = "none"
    provider_scenario: Literal[
        "succeeded",
        "failed",
        "processing_then_succeeded",
        "unknown_then_succeeded",
        "unknown_persistent",
    ] = "succeeded"
    idempotency_key: str = Field(min_length=8, max_length=100)
    approval_token: str | None = None

    @model_validator(mode="after")
    def validate_other_reason(self) -> "RefundExecuteRequest":
        if self.reason_code == "other" and len((self.reason_note or "").strip()) < 3:
            raise ValueError("reason_note is required when reason_code is other")
        return self


class RefundOperationActionRequest(BaseSchema):
    expected_version: int = Field(ge=1)
    idempotency_key: str = Field(min_length=8, max_length=100)


class RefundProviderWebhookRequest(BaseSchema):
    provider_event_id: str = Field(min_length=8, max_length=160)
    payment_leg_id: uuid.UUID
    sequence: int = Field(ge=1)
    state: Literal["processing", "succeeded", "failed", "unknown"]
    provider_refund_ref: str | None = Field(default=None, max_length=255)
    error_code: str | None = Field(default=None, max_length=80)


class RefundQuoteRead(BaseSchema):
    id: uuid.UUID
    order_id: uuid.UUID
    shift_id: uuid.UUID
    status: str
    currency: str
    order_version: int
    quote_hash: str
    items: list[dict[str, Any]]
    payment_allocations: list[dict[str, Any]]
    totals: dict[str, Any]
    policy: dict[str, Any]
    expires_at: datetime


class RefundOperationRead(BaseSchema):
    id: uuid.UUID
    order_id: uuid.UUID
    quote_id: uuid.UUID
    shift_id: uuid.UUID
    status: str
    reason_code: str
    reason_note: str | None
    currency: str
    subtotal_amount: Decimal
    discount_amount: Decimal
    vat_amount: Decimal
    rounding_amount: Decimal
    total_amount: Decimal
    stock_disposition: str
    provider_scenario: str
    row_version: int
    failure_code: str | None
    failure_message: str | None
    finalized_at: datetime | None
    tax_completed_at: datetime | None
    items: list[dict[str, Any]]
    payment_legs: list[dict[str, Any]]
    tax: dict[str, Any] | None
    created_at: datetime


class ReceiptData(BaseSchema):
    order: SaleOrderRead
    company_name: str
    branch_name: str
    cashier_name: str
    receipt_footer: str = "ขอบคุณที่ใช้บริการ"


class SyncSalesRequest(BaseSchema):
    orders: list[CreateSaleRequest] = Field(default_factory=list)


class HoldDraftItemRequest(BaseSchema):
    product_id: uuid.UUID
    variant_id: uuid.UUID | None = None
    qty: Decimal = Field(gt=0)
    discount_amount: Decimal = Field(default=Decimal("0"), ge=0)
    discount_type: Literal["amount", "percent"] = "amount"
    expected_unit_price: Decimal | None = Field(default=None, ge=0)
    expected_price_version: str | None = Field(default=None, min_length=1, max_length=128)
    price_override: PriceOverrideIntent | None = None
    display_name: str | None = Field(default=None, max_length=500)
    display_variant_name: str | None = Field(default=None, max_length=255)
    display_sku: str | None = Field(default=None, max_length=100)
    display_unit_code: str | None = Field(default=None, max_length=20)


class HoldDraftCreateRequest(BaseSchema):
    shift_id: uuid.UUID
    location_id: uuid.UUID
    label: str = Field(min_length=1, max_length=160)
    source_type: Literal["walk_in", "takeaway", "restaurant_table", "restaurant_quick_service"] = "walk_in"
    table_id: uuid.UUID | None = None
    queue_label: str | None = Field(default=None, max_length=80)
    customer_id: uuid.UUID | None = None
    customer_display: str | None = Field(default=None, max_length=160)
    note: str | None = Field(default=None, max_length=1000)
    items: list[HoldDraftItemRequest] = Field(min_length=1, max_length=100)
    order_discount: Decimal = Field(default=Decimal("0"), ge=0)
    loyalty_discount_intent: Decimal = Field(default=Decimal("0"), ge=0)
    currency: str = Field(default="THB", min_length=3, max_length=3)
    cart_version: int = Field(default=1, ge=1)
    idempotency_key: str = Field(min_length=8, max_length=100)


class HoldDraftUpdateRequest(BaseSchema):
    expected_version: int = Field(ge=1)
    idempotency_key: str = Field(min_length=8, max_length=100)
    label: str | None = Field(default=None, min_length=1, max_length=160)
    note: str | None = Field(default=None, max_length=1000)
    assignee_user_id: uuid.UUID | None = None
    reason: str | None = Field(default=None, max_length=500)
    items: list[HoldDraftItemRequest] | None = Field(default=None, min_length=1, max_length=100)
    order_discount: Decimal | None = Field(default=None, ge=0)
    loyalty_discount_intent: Decimal | None = Field(default=None, ge=0)
    cart_version: int | None = Field(default=None, ge=1)


class HoldDraftActionRequest(BaseSchema):
    expected_version: int = Field(ge=1)
    idempotency_key: str = Field(min_length=8, max_length=100)
    shift_id: uuid.UUID | None = None
    location_id: uuid.UUID | None = None
    claim_id: uuid.UUID | None = None
    accept_revalidation: bool = False
    reason: str | None = Field(default=None, max_length=500)


class HoldDraftRead(BaseSchema):
    id: uuid.UUID
    draft_no: str
    company_id: uuid.UUID
    brand_id: uuid.UUID | None = None
    branch_id: uuid.UUID
    location_id: uuid.UUID
    origin_shift_id: uuid.UUID
    owner_user_id: uuid.UUID
    owner_display: str | None = None
    assignee_user_id: uuid.UUID | None = None
    assignee_display: str | None = None
    origin_device_id: uuid.UUID | None = None
    origin_device_code: str | None = None
    origin_shift_number: str | None = None
    location_name: str | None = None
    parent_draft_id: uuid.UUID | None = None
    converted_order_id: uuid.UUID | None = None
    label: str
    source_type: str
    table_id: uuid.UUID | None = None
    queue_label: str | None = None
    customer_id: uuid.UUID | None = None
    customer_display: str | None = None
    note: str | None = None
    content: dict[str, Any]
    pricing_context: dict[str, Any]
    pricing_snapshot: dict[str, Any]
    last_revalidation: dict[str, Any] | None = None
    status: Literal["active", "claimed", "resumed", "expired", "converted", "cancelled"]
    version: int
    claim_id: uuid.UUID | None = None
    claimed_by: uuid.UUID | None = None
    claimed_device_id: uuid.UUID | None = None
    claim_expires_at: datetime | None = None
    expires_at: datetime
    resumed_at: datetime | None = None
    resumed_by: uuid.UUID | None = None
    expired_at: datetime | None = None
    cancelled_at: datetime | None = None
    cancel_reason: str | None = None
    converted_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class HoldDraftClaimRead(BaseSchema):
    draft: HoldDraftRead
    resume_cart: dict[str, Any]
    price_changes: list[dict[str, Any]] = Field(default_factory=list)
    requires_review: bool = False

File: frontend/src/types/pos.ts
export type PaymentMethod = "cash" | "promptpay" | "credit_card" | "bank_transfer" | "other";
export type ShiftStatus = "open" | "closed";
export type OrderStatus = "completed" | "voided" | "partially_refunded" | "refunded" | "pending_sync";

export interface CashierShift {
  id: string;
  shift_number: string;
  shift_type: "staff_cashier" | "operational_cashless";
  version: number;
  status: ShiftStatus;
  branch_id: string;
  location_id: string;
  user_id: string;
  opened_at: string;
  closed_at: string | null;
  opening_cash: number;
  closing_cash: number | null;
  expected_cash: number | null;
  cash_difference: number | null;
  total_sales: number;
  total_orders: number;
  total_voids: number;
  close_reason_code: string | null;
  cash_count_json: Array<{ denomination: number; quantity: number }> | null;
  opened_device_id: string | null;
  opened_device_code: string | null;
  closed_device_id: string | null;
  closed_device_code: string | null;
  closed_by_user_id: string | null;
  close_snapshot_json: PosShiftSummary | null;
}

export interface ShiftBlocker {
  code: string;
  count: number;
  message: string;
  action_path: string;
}

export interface ShiftCashMovement {
  id: string;
  movement_type: "cash_in" | "cash_out";
  amount: string;
  reason_code: string;
  reason: string;
  requester_id: string;
  approver_id: string | null;
  posted_at: string;
  journal_entry_id: string | null;
}

export interface PosShiftSummary {
  shift_id: string;
  shift_number: string;
  shift_type: string;
  status: ShiftStatus;
  version: number;
  company_id: string;
  branch_id: string;
  location_id: string;
  operator_user_id: string;
  opened_at: string;
  opening_cash: string;
  gross_sales: string;
  net_sales: string;
  refund_total: string;
  void_total: string;
  order_count: number;
  void_count: number;
  payment_totals: Record<string, string>;
  cash_received_net: string;
  change_total: string;
  cash_in_total: string;
  cash_out_total: string;
  expected_cash: string;
  cash_movements: ShiftCashMovement[];
  pending: {
    hold_drafts: number;
    refunds: number;
    offline_operations: number;
    unresolved_payments: number;
  };
  journal: {
    state: "matched" | "needs_reconciliation" | "not_applicable";
    missing_sales: number;
    missing_cash_movements: number;
  };
  blockers: ShiftBlocker[];
  can_close: boolean;
  reopen_allowed: false;
}

export interface CashMovement {
  id: string;
  shift_id: string;
  movement_type: "cash_in" | "cash_out";
  amount: number;
  reason_code: string;
  reason: string;
  status: "posted";
  shift_version: number;
  requester_id: string;
  approver_id: string | null;
  device_code: string | null;
  journal_entry_id: string | null;
  posted_at: string;
}

export interface CartItem {
  product_id: string;
  variant_id: string | null;
  product_name: string;
  variant_name: string | null;
  sku: string;
  unit_code: string | null;
  qty: number;
  unit_price: number;
  original_price: number;
  discount_amount: number;
  discount_type: "amount" | "percent";
  vat_type: string;
  vat_rate: number;
  subtotal: number;
  vat_amount: number;
  expected_price_version?: string | null;
  price_override?: {
    requested_unit_price: number;
    reason_code?: "customer_recovery" | "price_match" | "manager_comp" | "damaged_item" | "manual_correction" | "other";
    reason: string;
  } | null;
}

export interface Cart {
  items: CartItem[];
  discount_amount: number;
  discount_type: "amount" | "percent";
  subtotal: number;
  order_discount: number;
  vat_amount: number;
  total_amount: number;
}

export interface SaleOrderItem {
  id: string;
  product_id: string;
  variant_id: string | null;
  product_name: string;
  variant_name: string | null;
  sku: string;
  unit_code: string | null;
  qty: number;
  unit_price: number;
  original_price: number;
  discount_amount: number;
  vat_type: string;
  vat_rate: number;
  vat_amount: number;
  subtotal: number;
  refunded_qty?: number;
  refunded_amount?: number;
  price_source?: string | null;
  price_list_id?: string | null;
  price_list_version?: number | null;
  price_version?: string | null;
  price_snapshot?: Record<string, unknown> | null;
  order_discount_share?: number;
  line_total?: number;
  price_override_applied?: boolean;
  price_override_reason?: string | null;
}

export interface Payment {
  id: string;
  payment_method: PaymentMethod;
  amount: number;
  reference_no: string | null;
  original_payment_id?: string | null;
  currency?: string;
  provider_name?: string | null;
  provider_payment_ref?: string | null;
  settlement_state?: string;
  refund_operation_id?: string | null;
  provider_refund_state?: string | null;
  paid_at: string;
}

export type RefundOperationStatus =
  | "requested"
  | "processing"
  | "cash_due"
  | "succeeded"
  | "failed"
  | "unknown"
  | "needs_reconciliation"
  | "tax_pending"
  | "completed";

export interface RefundQuote {
  id: string;
  order_id: string;
  shift_id: string;
  status: string;
  currency: string;
  order_version: number;
  quote_hash: string;
  items: Array<{
    sale_order_item_id: string;
    product_name: string;
    quantity: string;
    subtotal_amount: string;
    discount_amount: string;
    vat_amount: string;
    total_amount: string;
    stock_disposition: "none" | "sellable";
  }>;
  payment_allocations: Array<{
    original_payment_id: string;
    payment_method: string;
    leg_type: "cash" | "provider";
    provider_name: string | null;
    amount: string;
    currency: string;
  }>;
  totals: {
    subtotal_amount: string;
    discount_amount: string;
    vat_amount: string;
    rounding_amount: string;
    total_amount: string;
    remaining_refundable_before: string;
    rounding_rule: string;
  };
  policy: Record<string, unknown>;
  expires_at: string;
}

export interface RefundOperation {
  id: string;
  order_id: string;
  quote_id: string;
  shift_id: string;
  status: RefundOperationStatus;
  reason_code: string;
  reason_note: string | null;
  currency: string;
  subtotal_amount: number;
  discount_amount: number;
  vat_amount: number;
  rounding_amount: number;
  total_amount: number;
  stock_disposition: string;
  provider_scenario: string;
  row_version: number;
  failure_code: string | null;
  failure_message: string | null;
  finalized_at: string | null;
  tax_completed_at: string | null;
  items: Array<Record<string, unknown>>;
  payment_legs: Array<{
    id: string;
    payment_method: string;
    leg_type: "cash" | "provider";
    provider_name: string | null;
    amount: number;
    currency: string;
    status: string;
    provider_refund_ref: string | null;
    attempt_count: number;
    last_error_code: string | null;
  }>;
  tax: {
    status: string;
    original_document_id: string | null;
    credit_note_id: string | null;
    retry_count: number;
    last_error: string | null;
  } | null;
  created_at: string;
}

export interface PaymentDraft {
  payment_method: PaymentMethod;
  amount: number;
  reference_no?: string | null;
}

export interface ExchangeContextDraft {
  source_order_id: string;
  source_order_number: string;
  refund_amount: number;
  refunded_items: string[];
  source_items?: Array<{
    product_id: string;
    product_name: string;
    qty?: number;
  }>;
  refund_reason: string;
}

export interface ReplacementRuleDraft {
  id: string;
  branch_id: string | null;
  source_product_id: string;
  source_product_name: string;
  replacement_product_id: string;
  replacement_product_name: string;
  created_at: number;
}

export interface SaleOrder {
  id: string;
  order_number: string;
  status: OrderStatus;
  branch_id: string;
  location_id: string;
  shift_id: string;
  user_id: string;
  customer_name: string | null;
  customer_phone: string | null;
  customer_tax_id?: string | null;
  subtotal: number;
  discount_amount: number;
  vat_amount: number;
  total_amount: number;
  refund_amount?: number;
  paid_amount: number;
  change_amount: number;
  is_offline: boolean;
  note: string | null;
  created_at: string;
  synced_at?: string | null;
  items: SaleOrderItem[];
  payments: Payment[];
  pricing_quote_id?: string | null;
  pricing_request_hash?: string | null;
  pricing_calculation_hash?: string | null;
  pricing_calculation_version?: string | null;
  pricing_context?: Record<string, unknown> | null;
  pricing_snapshot?: Record<string, unknown> | null;
  row_version?: number;
}

export interface PricingLineResult {
  product_id: string;
  variant_id: string | null;
  authoritative_unit_price: number;
  applied_unit_price: number;
  promotion_code?: string | null;
  promotion_discount_amount?: number;
  price_version: string;
  effective_at?: string;
  rounding_rule?: string;
  line_total: number;
  discrepancy: boolean;
  override_requested: boolean;
  override_requires_approval: boolean;
}

export interface PricingCalculation {
  quote_id: string;
  idempotency_key: string;
  calculation_hash: string;
  calculation_version: string;
  cart_version: number;
  subtotal: number;
  discount_amount: number;
  promotion_discount_amount?: number;
  discount_percentage: number;
  vat_amount: number;
  total_amount: number;
  rounding_rule?: string;
  effective_at?: string;
  expires_at: string;
  requires_price_override_approval: boolean;
  has_price_discrepancy: boolean;
  lines: PricingLineResult[];
}

export interface PendingSale {
  client_order_id: string;
  shift_id: string;
  location_id: string;
  items: CartItem[];
  customer_id?: string | null;
  discount_amount: number;
  discount_type: string;
  payment_method: PaymentMethod;
  payments?: PaymentDraft[];
  payment_reference?: string | null;
  paid_amount: number;
  customer_name: string | null;
  customer_phone?: string | null;
  customer_tax_id?: string | null;
  note?: string | null;
  exchange_context?: ExchangeContextDraft | null;
  total_amount: number;
  change_amount: number;
  created_at: number;
  synced: boolean;
  approval_token?: string | null;
}

export interface HeldSaleDraft {
  id: string;
  shift_id: string;
  location_id: string;
  branch_id: string | null;
  sales_channel?: "walk_in" | "takeaway";
  label: string;
  items: CartItem[];
  order_discount: number;
  loyalty_discount: number;
  payment_method: PaymentMethod;
  paid_amount: number;
  payment_reference: string | null;
  split_payment_enabled?: boolean;
  secondary_payment_method?: PaymentMethod;
  secondary_payment_amount?: number;
  secondary_payment_reference?: string | null;
  customer_name: string;
  customer_phone: string;
  customer_tax_id: string;
  customer_search: string;
  selected_customer: import("@/types/crm").Customer | null;
  note: string;
  exchange_context?: ExchangeContextDraft | null;
  held_at: number;
  sync_state?: "local_only" | "pending_sync" | "synced" | "needs_review";
  server_id?: string;
  draft_no?: string;
  status?: "active" | "claimed" | "resumed" | "expired" | "converted" | "cancelled";
  version?: number;
  expires_at?: string;
  owner_user_id?: string;
  owner_display?: string | null;
  assignee_user_id?: string | null;
  assignee_display?: string | null;
  origin_device_id?: string | null;
  origin_device_code?: string | null;
  origin_shift_number?: string | null;
  location_name?: string | null;
  table_id?: string | null;
  queue_label?: string | null;
  claim_id?: string | null;
  claimed_by?: string | null;
  claim_expires_at?: string | null;
  last_revalidation?: Record<string, unknown> | null;
  resumed_at?: string | null;
  expired_at?: string | null;
  cancelled_at?: string | null;
  converted_at?: string | null;
  cancel_reason?: string | null;
  server_backed?: boolean;
}

export interface ServerHoldDraft {
  id: string;
  draft_no: string;
  company_id: string;
  brand_id: string | null;
  branch_id: string;
  location_id: string;
  origin_shift_id: string;
  owner_user_id: string;
  owner_display: string | null;
  assignee_user_id: string | null;
  assignee_display: string | null;
  origin_device_id: string | null;
  origin_device_code: string | null;
  origin_shift_number: string | null;
  location_name: string | null;
  parent_draft_id: string | null;
  converted_order_id: string | null;
  label: string;
  source_type: "walk_in" | "takeaway" | "restaurant_table" | "restaurant_quick_service";
  table_id: string | null;
  queue_label: string | null;
  customer_id: string | null;
  customer_display: string | null;
  note: string | null;
  content: {
    items: CartItem[];
    order_discount: string;
    loyalty_discount_intent: string;
    currency: string;
    cart_version: number;
  };
  pricing_context: Record<string, unknown>;
  pricing_snapshot: Record<string, unknown>;
  last_revalidation: Record<string, unknown> | null;
  status: "active" | "claimed" | "resumed" | "expired" | "converted" | "cancelled";
  version: number;
  claim_id: string | null;
  claimed_by: string | null;
  claimed_device_id: string | null;
  claim_expires_at: string | null;
  expires_at: string;
  resumed_at: string | null;
  resumed_by: string | null;
  expired_at: string | null;
  cancelled_at: string | null;
  cancel_reason: string | null;
  converted_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface HoldDraftAuditEntry {
  id: string;
  action: string;
  from_status: ServerHoldDraft["status"] | null;
  to_status: ServerHoldDraft["status"];
  from_version: number | null;
  to_version: number;
  actor_user_id: string;
  actor_display: string | null;
  device_id: string | null;
  shift_id: string | null;
  reason: string | null;
  metadata: Record<string, unknown>;
  created_at: string;
}

export interface HoldDraftClaimResult {
  draft: ServerHoldDraft;
  resume_cart: {
    items: CartItem[];
    pricing: PricingCalculation & Record<string, unknown>;
  };
  price_changes: Array<Record<string, unknown>>;
  requires_review: boolean;
}
